src/instructlab/model/phased_training.py

```diff
@@ -0,0 +1,182 @@
+    score: float
+
+    @pydantic.field_serializer("checkpoint", "ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class EvalPhaseModel(pydantic.BaseModel):
+    """Stores info about evaluation phase"""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: list[pydantic.DirectoryPath]
+    finished_checkpoints: list[pydantic.DirectoryPath] = []
+    results: list[EvalResult] = []
+    best_checkpoint: EvalResult | None = None
+
+    @pydantic.field_serializer(
+        "checkpoints",
+        "finished_checkpoints",
+    )
+    def pathlibPath_list_to_str(self, paths: list[pathlib.Path]) -> list[str]:
+        return [str(path) for path in paths]
+
+    @pydantic.field_serializer("ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class TrainPhaseModel(pydantic.BaseModel):
+    """Stores info about training phase. Doesn't store training args because user can get those elsewhere, e.g. config."""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: pydantic.DirectoryPath
+    # TODO: might want training args here, but can't currently do it
+    # without side-effects because model instance is changed inside of _training_phase
+
+    @pydantic.field_serializer("checkpoints", "ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class JournalModel(pydantic.BaseModel):
+    run_id: uuid.UUID = pydantic.Field(default_factory=uuid.uuid4)
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    current_phase: TrainingPhases = TrainingPhases.TRAIN1
+    train_1: TrainPhaseModel | None = None
+    eval_1: EvalPhaseModel | None = None
+    train_2: TrainPhaseModel | None = None
+    eval_2: EvalPhaseModel | None = None
+    final_output: EvalResult | None = None
+
+    @pydantic.field_serializer("current_phase")
+    def enum_to_value(self, enum_val: TrainingPhases) -> str:
+        return str(enum_val.value)
+
+
+class TrainingJournal:
+    def __init__(self, journalfile: pathlib.Path):
+        self.journalfile = journalfile
+
+        if journalfile.is_dir() or (not journalfile.exists()):  # otherwise, is a file.
+            raise ValueError(
+                f"Path to 'journalfile.yaml' must exist and be to a .yaml file: {str(journalfile)}"
+            )
+
+        valid_extentions = [".yaml", ".yml"]
+        _, file_extension = os.path.splitext(str(journalfile))
+        if file_extension not in valid_extentions:
+            raise ValueError(
+                f"Journalfile does not have a valid file extention.\nExtension: {file_extension}\nValid extensions: {valid_extentions}"
+            )
+
+        # is a file, can be empty
+        if journalfile.stat().st_size == 0:
```

Comment:
I don't really see the value in comments / whitespace handling since this is an internal bookkeeping document that users aren't supposed to routinely modify.